coderd/userauth.go

```diff
@@ -205,3 +209,137 @@ func (api *API) userOAuth2Github(rw http.ResponseWriter, r *http.Request) {
 	}
 	http.Redirect(rw, r, redirect, http.StatusTemporaryRedirect)
 }
+
+type OIDCConfig struct {
+	httpmw.OAuth2Config
+
+	Verifier *oidc.IDTokenVerifier
+	// EmailDomain is the domain to enforce when a user authenticates.
+	EmailDomain  string
+	AllowSignups bool
+}
+
+func (api *API) userOIDC(rw http.ResponseWriter, r *http.Request) {
+	state := httpmw.OAuth2(r)
+
+	// See the example here: https://github.com/coreos/go-oidc
+	rawIDToken, ok := state.Token.Extra("id_token").(string)
+	if !ok {
+		httpapi.Write(rw, http.StatusBadRequest, codersdk.Response{
+			Message: "id_token not found in response payload. Ensure your OIDC callback is configured correctly!",
+		})
+		return
+	}
+
+	idToken, err := api.OIDCConfig.Verifier.Verify(r.Context(), rawIDToken)
+	if err != nil {
+		httpapi.Write(rw, http.StatusBadRequest, codersdk.Response{
+			Message: "Failed to verify OIDC token.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+
+	var claims struct {
+		Email    string `json:"email"`
+		Verified bool   `json:"email_verified"`
+		Username string `json:"preferred_username"`
+	}
+	err = idToken.Claims(&claims)
+	if err != nil {
+		httpapi.Write(rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Failed to extract OIDC claims.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+	if claims.Email == "" {
+		httpapi.Write(rw, http.StatusBadRequest, codersdk.Response{
+			Message: "No email found in OIDC payload!",
+		})
+		return
+	}
+	if !claims.Verified {
+		httpapi.Write(rw, http.StatusForbidden, codersdk.Response{
+			Message: fmt.Sprintf("Verify the %q email address on your OIDC provider to authenticate!", claims.Email),
+		})
+		return
+	}
+	// The username is a required property in Coder. We make a best-effort
+	// attempt at using what the claims provide, but if that fails we will
+	// generate a random username.
+	if !username.Valid(claims.Username) {
+		// If no username is provided, we can default to use the email address.
+		// This will be converted in the from function below, so it's safe
+		// to keep the domain.
+		if claims.Username == "" {
+			claims.Username = claims.Email
+		}
+		claims.Username = username.From(claims.Username)
+	}
+	if api.OIDCConfig.EmailDomain != "" {
+		if !strings.HasSuffix(claims.Email, api.OIDCConfig.EmailDomain) {
+			httpapi.Write(rw, http.StatusForbidden, codersdk.Response{
+				Message: fmt.Sprintf("Your email %q is not a part of the %q domain!", claims.Email, api.OIDCConfig.EmailDomain),
+			})
+			return
+		}
+	}
+
+	var user database.User
+	user, err = api.Database.GetUserByEmailOrUsername(r.Context(), database.GetUserByEmailOrUsernameParams{
```

Comment:
This is pretty unlikely to happen but if an existing, unrelated user (say with `build-in` auth) already has this email address, then it's possible for a new OIDC user to impersonate them by providing the same email address.

Reply:
Yup. The issue I attached expands on that a bit. We'll have to make the user-identification with account links better.

Reply:
Do we want to add a column to the users table that indicates the sort of authentication they used? Probably not necessary to do in this PR

Reply:
We will in a future PR, this one just gets it out the door